Initial implementation

mattruggio opened the pull request:
This PR is now ready for review.

gerryster commented on lib/proforma/plain_text_renderer.rb:
```diff
@@ -0,0 +1,106 @@
+module Proforma
+  # A basic rendering engine that will output plain-text.  It is meant to serve as
+  # an example of how to create a rendering engine for this library.
+  class PlainTextRenderer
```
Nice! I like that a simple render can be implemented in about 100 lines of code.

gerryster commented on spec/proforma_spec.rb:
```diff
@@ -0,0 +1,230 @@
+
+      expected_documents = [
+        Proforma::Document.new(
+          contents: "========================================\nSystem A\n=================="\
```
How about defining the contents in a here document for better readability?

mattruggio replied:
Agreed.

gerryster replied:
The fixture is even better.

gerryster commented on spec/spec_helper.rb:
```diff
@@ -7,11 +7,19 @@
+
+def yaml_read(file)
+  # rubocop:disable Security/YAMLLoad
+  YAML.load(File.open(file))
```
Can this use safe_load instead?

mattruggio replied:
Technically in this library yes.  I may find use of unsafe_load in other proforma libraries where I want to store/snapshot non-hash-based Ruby objects.

mattruggio commented on README.md:
````diff
@@ -1,5 +1,328 @@
+The basic evaluator that comes with this library is Proforma::HashEvaluator.  It is limited to:
+
+1. only resolving values for hash objects and cannot handle nesting / recursion / dot notation.
+2. only text templating simple expressions: if the expression starts with a ```$:``` then it is noted as being a data-bound value.  The value proceeding the ```$:``` will be resolved and returned.  If it does not begin with this special string then it is deemed just a basic string and it will be returned verbatim.
````
That sounds better.

mattruggio commented on README.md:
```diff
@@ -1,5 +1,328 @@
+
+Other libraries that can be plugged in, such as:
+
+1. [![proforma-extended-evaluator](https://github.com/bluemarblepayroll/proforma-extended-evaluator): adds nested dot-notation object value resolution and rich text templating.
```
Those were malformed links.

mattruggio commented on bin/render:
```diff
@@ -0,0 +1,68 @@
+
+def read_yaml(filename)
+  # rubocop:disable Security/YAMLLoad
+  YAML.load(read(filename))
```
Technically, in this repository, yes since the data is hash-based and the template has a hash interface.  The original intent was to allow YAML to marshal Ruby objects straight from YAML but since has been removed from this repo.  Other repos, such as the extended-evaluator and possibly the prawn-renderer may make use of this since some of their snapshot Ruby objects do not have a hash interface.

mattruggio commented on lib/proforma/compiling/aggregation.rb:
```diff
@@ -0,0 +1,62 @@
+        self
+      end
+
+      def to_hash
```
No, good call.

mattruggio commented on lib/proforma/compiling/counter.rb:
```diff
@@ -0,0 +1,35 @@
+      end
+
+      def ave
+        sum / count
```
Agreed.

mattruggio commented on lib/proforma/hash_evaluator.rb:
```diff
@@ -0,0 +1,40 @@
+  # - if text is prefixed with a dollar sign and colon then it means it will be dynamically
+  #   evaluated against the record.  For example: $:id
+  class HashEvaluator
+    PROPERTY_PREFIX = '$:'
```
no!  For even more simplicity I will switch to just a dollar sign.  In reality a different evaluator will be used, this merely serves as a baseline implementation.

gerryster replied:
My inner Perl programmer is very satisfied.

mattruggio commented on lib/proforma/hash_evaluator.rb:
```diff
@@ -0,0 +1,40 @@
+      return object if expression.to_s.empty?
+      return nil    unless object.is_a?(Hash)
+
+      if object.key?(expression.to_s)
```
Technically they are different:

Yours: `{ a: 1, 'a' => nil } # would resolve to: 1`
Mine: `{ a: 1, 'a' => nil } # would resolve to: nil`

For a baseline implementation of a resolver, this argument is rather pedantic.  I would be fine simplifying it to your example.

mattruggio commented on lib/proforma/modeling.rb:
```diff
@@ -0,0 +1,21 @@
+# LICENSE file in the root directory of this source tree.
+#
+
+require_relative 'modeling/generic_container'
```
I am torn between either explicitly just requiring it within each file or simplifying it and only requiring it here.  For this case I will just add a blank line.

mattruggio commented on lib/proforma/modeling/generic_container.rb:
```diff
@@ -0,0 +1,57 @@
+
+        args = {}.tap { |hash| hash[child_key] = compiled_children }
+
+        self.class.new(args)
```
No I think you are right.  This should be a good simplification.

mattruggio commented:
@gerryster - I addressed all feedback.  Even if I did not comment on each individual one I believe I agreed with all feedback and, as such, there should be appropriate changes.

gerryster commented on README.md:
```diff
@@ -1,5 +1,328 @@
+
+The `contents` attribute will now contain:
+
+* banner
```
Cool, thanks for adding a link to the fixture file!

gerryster commented:
This looks good. Just one more item not responded to which is the suggestion about refinements for symbolize_keys instead of monkey patching. I guess the concern here would be if some other gem monkey patches an identically named method in a slightly different way. Probably not a big concern but refinements would make it a non-issue.

mattruggio commented:
@gerryster - I updated the monkey patch to a refinement, totally missed that one piece of feedback.  Ready for another look.

gerryster commented:
Nice the refinement looks good!

gerryster commented:
LGTM
